### PR TITLE
 Add allowTransparency option

### LINKED ISSUE #1004: Support transparent backgrounds
It looks as though starting with 365661704424f9a33b3a2cb4b85ad144948a86a9 setting the themes background to 'none' does not work. It just sets the background to black. If it helps at all I found changing this line: https://github.com/sourcelair/xterm.js/blob/c4da8325ace6e4bde77822377bf994d389a7c75a/src/renderer/TextRenderLayer.ts#L104 to `let bg = this._colors.background` fixed the issue but not sure if that is the correct solution to the problem.

### PR DESCRIPTION
**Pull request by chabou**
Fixes #1004 (and replaces #1148)

I didn't notice any performance loss (measuring frame rendering duration for vtop on electron).

### CODE REVIEW

File: src/renderer/CharAtlas.ts [hunk 1]
```diff
@@ -116,19 +116,19 @@ let generator: CharAtlasGenerator;
  * Initializes the char atlas generator.
  * @param document The document.
  */
-export function initialize(document: Document): void {
+export function initialize(document: Document, alpha: boolean): void {
```
Comment (Tyriar) [inline src/renderer/CharAtlas.ts]: Looks like `setOption('allowTransparency', true)` will not work when set after the page's first `new Terminal()` call. We need a handler in `setOption` and for `CharAtlas.ts` to regenerate the canvas when `alpha` changes.

File: src/renderer/TextRenderLayer.ts [hunk 2]
```diff
@@ -24,8 +24,8 @@ export class TextRenderLayer extends BaseRenderLayer {
   private _characterFont: string;
   private _characterOverlapCache: { [key: string]: boolean } = {};
 
-  constructor(container: HTMLElement, zIndex: number, colors: IColorSet) {
-    super(container, 'text', zIndex, false, colors);
+  constructor(container: HTMLElement, zIndex: number, colors: IColorSet, alpha: boolean) {
+    super(container, 'text', zIndex, alpha, colors);
```
Comment (Tyriar) [inline src/renderer/TextRenderLayer.ts]: Same here, changing after the first `new Terminal()`will not affect `TextRenderLayer`

File: typings/xterm.d.ts [hunk 3]
```diff
@@ -409,7 +415,7 @@ declare module 'xterm' {
      * Retrieves an option's value from the terminal.
      * @param key The option key.
      */
-    getOption(key: 'cancelEvents' | 'convertEol' | 'cursorBlink' | 'debug' | 'disableStdin' | 'enableBold' | 'popOnBell' | 'screenKeys' | 'useFlowControl' | 'visualBell'): boolean;
+    getOption(key: 'cancelEvents' | 'convertEol' | 'cursorBlink' | 'debug' | 'disableStdin' | 'enableBold' | 'popOnBell' | 'screenKeys' | 'useFlowControl' | 'visualBell' | 'allowTransparency'): boolean;
```
Comment (Tyriar) [inline typings/xterm.d.ts]: Let's put the new option at the start

File: src/renderer/CharAtlas.ts (hunk 1 above)
Comment (chabou) [reply src/renderer/CharAtlas.ts]: Will do that

**Comment by Tyriar** [comment]
@chabou I resolved the conflicts, I also fixed the canvas alpha updating after the option changed when *requesting* the char atlas as there was some refactoring done that made that trivial, as well as the typing ordering.

Changes still needed:

- Make sure `setOption('allowTransparency', ...)` requests a new char atlas and updates the `TextRenderLayer` canvas properly

**Comment by chabou** [comment]
@Tyriar My last commit https://github.com/xtermjs/xterm.js/pull/1236/commits/fe53e1a09136828cb32faab5326d48ec11456c1f clones canvas to obtain a new context with desired transparency capability.
This change needs to be done in BaseRenderer class (to access some private properties/methods), but only TextRenderer need to recreate its canvas and regenerate its char atlas.

I am not sure that is the best way to handle this and I'll be happy to update it to suit your needs.

**Comment by Tyriar** [comment]
I made a few minor changes and it looks good. I removed the call to `this._ctx.scale` as it was breaking the scale of the terminal after `setOption('allowTransparency', true)` was called. I'm not sure it was doing anything anyway.

**Comment by Tyriar** [comment]
🎉

File: src/renderer/BaseRenderLayer.ts [hunk 4]
```diff
@@ -54,23 +57,22 @@ export abstract class BaseRenderLayer implements IRenderLayer {
...
       return;
     }
-    this._alpha = alpha
-    const oldCanvas = this._canvas
-    // Cloning preserves canvas properties and permits to acquire a new context 
+
+    // Create new canvas and replace old one
+    const oldCanvas = this._canvas;
+    this._alpha = alpha;
+    // Closing preserves properties
```
Comment (chabou) [inline src/renderer/BaseRenderLayer.ts]: Litlle typo here: Closing -> Cloning

Comment (Tyriar) [reply src/renderer/BaseRenderLayer.ts]: Whoops 😛

Comment (Tyriar) [reply src/renderer/BaseRenderLayer.ts]: Fixed https://github.com/xtermjs/xterm.js/pull/1260

**Pull request merged**